### PR TITLE
Add HoaConsole readline implementation

### PR DESCRIPTION
**Pull request by zonuexe**
[hoaproject/Console: The Hoa\Console library.](https://github.com/hoaproject/Console)

Pure PHP readline implementation.

### CODE REVIEW

**Comment by zonuexe** [comment]
I made [zonuexe/psysh-hoa-console](https://github.com/BaguettePHP/psysh-hoa-console) package.

I hope to shrink the package by this pull-request is merged.

**Comment by bobthecow** [comment]
Thank you for this!

I'm a bit reluctant to add `hoa/console` to suggested dependencies, at least with the current text, because most(?) people will not need it. Should it be a runtime warning instead? Maybe more descriptive text would help?

> A pure PHP readline implementation. You'll want this if your PHP install doesn't already support readline or libedit.

But a suggested dependency doesn't help people who use the phar-based install :(

For the phar-based install, I could see building Hoa\Console into the phar by default. It looks like it would increase the size of the file by about 25-30%. I just pushed a commit that offsets it a bit by making the existing phar lighter by 10%, but that's still a non-trivial amount of code to add to support a small subset of users.

Alternatively, we could do what we do when the user is missing `ext-pdo-sqlite`: add it to composer suggested dependencies, but also add a runtime warning for users who run into it with a phar-based install.

#297 has a similar issue, but with `ext-iconv`.

Probably the best thing to do is what was suggested there: to add a more descriptive suggested dependency message to keep users from unintentionally installing this, and also to ship a second (bloated) phar. That phar would include Hoa\Console as a readline polyfill, Symfony's iconv polyfill, and anything else we come across like this.

How's that sound?

File: src/Psy/Readline/HoaConsole.php [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+use Hoa\Console\Readline\Readline as HoaReadline;
+use Psy\Exception\BreakException;
+
+/**
+ * Hoa\Console Readline implementation.
+ */
+class HoaConsole implements Readline
+{
+    /** @var HoaReadline */
+    private $hoa_readline;
```
Comment (bobthecow) [inline src/Psy/Readline/HoaConsole.php]: Let's call this `$hoaReadline` for consistency with the rest of the codebase.

File: src/Psy/Readline/HoaConsole.php [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    public static function isSupported()
+    {
+        return class_exists('\Hoa\Console\Console', true);
+    }
+
+    public function __construct()
+    {
+        $this->hoa_readline = new HoaReadline();
+        $this->reflection = new \ReflectionProperty($this->hoa_readline, '_history');
```
Comment (bobthecow) [inline src/Psy/Readline/HoaConsole.php]: Can we poke Hoa to get a proper API for history access? I'd rather not depend on reflection on a private member :(

File: src/Psy/Readline/HoaConsole.php [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+     * Get a STDIN file handle.
+     *
+     * @throws BreakException if user hits Ctrl+D
+     *
+     * @return resource
+     */
+    private function getStdin()
+    {
+        return Console::getInput()->getStream()->getStream();
+    }
```
Comment (bobthecow) [inline src/Psy/Readline/HoaConsole.php]: This looks unused?

File: src/Psy/Readline/HoaConsole.php (hunk 2 above)
Comment (Hywan) [inline src/Psy/Readline/HoaConsole.php]: What would you like?

**Comment by Hywan** [comment]
Hello 😃,
I am glad to see this PR! Several people asked for a `Hoa\Console` integration inside PsySH. Goals are numerous. `Hoa\Console` is not only a “pure readline implementation” as the description given in this PR suggests, but it is based on terminfo/tput, which makes it interoperable with most of the platforms out there. And regarding readline features, the autocomplete API is powerful and some default autocompleters are already present, like a `Word` or `Path`.
demo

I know some people have developed a shell with `Hoa\Console\Readline` (like https://github.com/jubianchi/atoum-shell cc @jubianchi) if they can help!

File: composer.json [hunk 4]
```diff
@@ -24,13 +24,15 @@
         "phpunit/phpunit": "~3.7|~4.0|~5.0",
         "symfony/finder": "~2.1|~3.0",
         "squizlabs/php_codesniffer": "~2.0",
-        "fabpot/php-cs-fixer": "~1.5"
+        "fabpot/php-cs-fixer": "~1.5",
+        "hoa/console": "~3.16|~1.14"
```
Comment (Hywan) [inline composer.json]: Why `~1.14`?

Comment (zonuexe) [inline composer.json]: This dependency is added for to support PHP 5.4 and 5.4.
https://travis-ci.org/bobthecow/psysh/builds/122592026

File: src/Psy/Readline/HoaConsole.php (hunk 2 above)
Comment (bobthecow) [inline src/Psy/Readline/HoaConsole.php]: Looks like we didn't need it. Thanks @zonuexe!

**Comment by zonuexe** [comment]
@bobthecow @Hywan 
Thank you for your review!

**Comment by bobthecow** [comment]
Thanks @zonuexe!

**Comment by bobthecow** [comment]
@Hywan any sense for how much work it'd be to get Hoa\Console autocomplete working with PsySH's current autocompletion? That'd be nice to support as well :)

**Comment by jubianchi** [comment]
this is really great! feel free to ping me for any help.

**Comment by Hywan** [comment]
@bobthecow I have no idea. I don't know how PsySH works. Maybe @zonuexe can give a try. What actually works so far? Is this PR provide readline support with `Hoa\Console` already?

File: src/Psy/Readline/HoaConsole.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        return $this->hoaReadline->readLine($prompt);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function redisplay()
+    {
+        // noop
```
Comment (Hywan) [inline src/Psy/Readline/HoaConsole.php]: @zonuexe, @bobthecow Actually this is possible to do that. What is the goal of this method exactly?

File: src/Psy/Readline/HoaConsole.php [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function listHistory()
+    {
+        $i = 0;
+        $list = array();
+        while (($item = $this->hoaReadline->getHistory($i++)) !== null) {
```
Comment (Hywan) [inline src/Psy/Readline/HoaConsole.php]: @zonuexe @bobthecow We can add a `getWholeHistory`-like method if needed. Thoughts?

Comment (bobthecow) [inline src/Psy/Readline/HoaConsole.php]: It'd be cleaner than this :)

Unfortunately, we need to support all the way back to PHP 5.3, so unless it's ported to the legacy version as well, it won't do us a ton of good.

File: src/Psy/Readline/HoaConsole.php (hunk 5 above)
Comment (bobthecow) [inline src/Psy/Readline/HoaConsole.php]: `Psy\Readline` is an abstraction over the whole readline API. But `redisplay` isn't _currently_ used in PsySH, so it's more there for completeness than anything else. If Hoa\Console supports it and it's trivial, we should add it for if/whenever PsySH needs to use that.

**Comment by bobthecow** [comment]
@Hywan PsySH currently has readline-based completion support, provided by [AutoCompleter](https://github.com/bobthecow/psysh/blob/master/src/Psy/TabCompletion/AutoCompleter.php). It's not using the Psy\Readline abstraction, but could be changed to use that. If that were the case, it _should_ be as simple as implementing `readline_completion_function` and `readline_info` as part of the Psy\Readline interface.

File: src/Psy/Readline/HoaConsole.php (hunk 5 above)
Comment (Hywan) [inline src/Psy/Readline/HoaConsole.php]: We can still clear the line and print it again. cc @zonuexe

**Comment by Hywan** [comment]
@bobthecow It's left to @zonuexe to decide whether she/he would like to implement autocompleter too. They are based on the readline though.

File: src/Psy/Readline/HoaConsole.php (hunk 5 above)
Comment (bobthecow) [inline src/Psy/Readline/HoaConsole.php]: I wouldn't worry about it, personally. it's not really a big deal, since nothing uses it :)

**Pull request merged**